### PR TITLE
ext_authz: Add support for headers_to_remove

### PR DESCRIPTION
**Pull request by numerodix**
Commit Message: ext_authz: Add support for headers_to_remove
Additional Description: Add support for the letting the authorization service tell Envoy which auth related headers to remove once the authorization server is done with them, so that the upstream does not see them.
Risk Level: Low
Testing: New tests added + integration tested manually using a simple mock auth service
Docs Changes: added
Release Notes: added

Signed-off-by: Martin Matusiak <numerodix@gmail.com>

----

ORIGINAL DESCRIPTION WHILE THIS PR WAS IN DRAFT MODE:

@dio I want to add a way for the authorization server to tell envoy "this particular header was meant for me, remove it before you forward the request to the upstream". 

However, the behavior is not meant to be hardcoded (otherwise it could be done in the envoy config directly using `request_headers_to_remove`). It's meant to be dynamic where the authorization server can decide which headers to look at, which ones it will "consume" for the purpose of authorization and which ones it will leave alone.

The greater use case here is "an authorization server that supports multiple auth mechanisms (that may use different sets of headers) and that can run under different configurations (where different auth methods may be enabled/disabled)".

So for example if the authorization server supports METHOD-X and METHOD-Y and the request contains:
```
HEADER-FOR-METHOD-X: something
HEADER-FOR-METHOD-Y: something-else
```

It can authorize based on `HEADER-FOR-METHOD-X` alone (if that is tried first, according to its config), tell envoy to remove it from the forwarded request, and leave `HEADER-FOR-METHOD-Y` as is. Or it can tell envoy to remove both, even if only one was needed for auth, all depending on how the auth server is configured to run.

I hope that makes the use case understandable. :) 

This feels straightforward to implement for the grpc client where we can add the protobuf field `headers_to_remove` that is part of the `OkResponse`. I've done that in this PR.

It's a bit more awkward to implement it in the http client, where all the communication is header based. We'd need to introduce a response header set by the auth server called something like `x-envoy-auth-server-headers-to-remove: a-header, another-header` and special case how envoy treats it (ie. the envoy should read the header, act on it, then remove it before forwarding the request).

Incidentally, I only need it in the grpc client, but I don't know if it's acceptable to have this lack of parity between the clients (?)

I'd very much appreciate your guidance on whether this looks sane to you, and what changes you think are needed!

### CODE REVIEW

**Comment by dio** [comment]
I think the motivation is reasonable. Cc. @htuch to get opinion on this. Thanks.

**Comment by numerodix** [comment]
> I think the underlying motivation is reasonable. While probably today we can set those headers as empty or meaningless headers..

@dio @htuch  The http authorization server would be responsible for setting them. We'd have to document exactly what they need to do, given that it's not part of a protobuf contract. And it would be optional, of course.

**Comment by htuch** [comment]
Seems reasonable to me. For HTTP ext_authz, I'd try and follow whatever pattern already exists for OOB comms.

**Comment by dio** [comment]
Great. Please let me know when this is ready for review. Thanks!

**Comment by numerodix** [comment]
> Great. Please let me know when this is ready for review. Thanks!

@dio Ready for your review now.

File: api/envoy/service/auth/v3/external_auth.proto
Comment (htuch) [inline api/envoy/service/auth/v3/external_auth.proto:84]: The other case I can see where headers are used is `x-envoy-auth-partial-body`. This is used for Envoy --> upstream, which is different, but it's a little weird to refer to the filter as auth vs. authz.

Comment (numerodix) [reply api/envoy/service/auth/v3/external_auth.proto]: Sure, let me make that consistent.

**Comment by numerodix** [comment]
@dio? :)

File: docs/root/version_history/current.rst
Comment (dio) [inline docs/root/version_history/current.rst:77]: nit: s/envoy/Envoy

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz.h:63]: Can we use `Http::HeaderVector` here?

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz.h:61]: Sorry, what do you mean by consumed by the authorization server?

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:328]: nit: 

```c++
StringUtil::splitToken(storage_header_value, ",", /*keep_empty_string=*/false, /*trim_whitespace=*/true)
```

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:322]: HeaderVector? Probably we don't need to initialize this here? I mean, we should do this only when we're sure that we have `EnvoyAuthHeadersToRemove` and  it is an OK response.

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (dio) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:328]: And it seems like before pushing the headers, we can allocate (`reserve`) the size of `headers_to_remove` here?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:169]: Sorry why we need this? Could you give some inline comments here?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: Do you think we need to have a ban list of headers that can't be removed?

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (dio) [inline test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:93]: "headers to be removed"?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (dio) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:1270]: HeaderVector?

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (dio) [inline test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:428]: s/envoy/Envoy

**Review by dio — CHANGES_REQUESTED** [review]
Flushing some comments on the implementations.

File: docs/root/version_history/current.rst
Comment (numerodix) [reply docs/root/version_history/current.rst:77]: Fixing

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (numerodix) [reply test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:428]: Fixing

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz.h:63]: HeaderVector is a `std::vector<std::pair<LowerCaseString, std::string>>`. This is perfect for `headers_to_append`, `headers_to_set` and `headers_to_add`, because those need both a key and a value. `headers_to_remove` only needs the header name (ie. the key). The value would always be empty. So I figured a plain `std::vector<Envoy::Http::LowerCaseString>` would be a better fit.

Happy to change this if you prefer `HeaderVector`?

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz.h:61]: Consumed in the sense that the consumer of these headers (ie. who these headers are for) is the authorization server. Would you like me to rephrase this?

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:328]: Fixing both of these.

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:322]: Fixing to avoid initializing.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (numerodix) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:169]: Yeah great point. I'll add a comment here.

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (dio) [reply source/extensions/filters/common/ext_authz/ext_authz.h:63]: Got it. I think we can keep this as is.

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (dio) [reply source/extensions/filters/common/ext_authz/ext_authz.h:61]: After reading this once again. Seems like it is fine. But probably this "(using "remove")" is not necessary?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (dio) [reply test/extensions/filters/http/ext_authz/ext_authz_test.cc:1270]: We can keep this as is.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: That's a very good question.

I assume that since we are giving the authorization server permission to change headers that means the authorization server is trusted by Envoy. So we are not concerned that it may act maliciously. But that still leaves the possibility that the authorization server may change the headers in such a way as to make the request malformed.

Perhaps we should prevent it from removing the headers:
* :method
* :authority
* :path
* ... any others?

What do you think?

I just tested this manually and removing either of `:method` or `:authority` makes Envoy crash. `:path` does not.

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (numerodix) [reply test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:93]: Fixing

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz.h:61]: Okay, I will update that.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: We could potentially also disallow removing:
* :protocol (could it be present in a request?)
* :scheme (could it be present in a request?)
* :status (should never be present in a request?)

Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: We could potentially also disallow removing:

* :protocol (could it be present in a request?)
* :scheme (could it be present in a request?)
* :status (should never be present in a request?)

Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: I found a relevant test for this:
https://github.com/envoyproxy/envoy/blob/master/test/common/router/config_impl_test.cc#L1723

Perhaps this is the list of headers we should disallow?

Comment (dio) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: Yeap. I think we should ban removing pseudo headers.

Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:62]: Implemented + added a new test for this.

**Comment by numerodix** [comment]
@dio are you happy with the changes now?
@htuch will need your api approval again after changing the docstring.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:210]: This works, but let's use `absl::StartsWithIgnoreCase` instead?

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (dio) [inline test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:473]: Seems like we haven't test it for multiple header names and its variations? 

1. `{"x-envoy-auth-headers-to-remove", "remove-1,remove-1"}`
1. `{"x-envoy-auth-headers-to-remove", "remove-1, remove-2"}`
1. `{"x-envoy-auth-headers-to-remove", "remove-1,,,remove-2"}`

And also when we have multiple entries of "x-envoy-auth-headers-to-remove".

```
{"x-envoy-auth-headers-to-remove", "remove-1,remove-2"}
{"x-envoy-auth-headers-to-remove", "remove-2,remove-3"}
```

Comment (dio) [reply test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:473]: Probably you can do this in unit tests below.

**Review by dio — CHANGES_REQUESTED** [review]
Awesome.

A small request for adding tests against malformed comma-separated header value.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (numerodix) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Fixed

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (numerodix) [reply test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:473]: @dio Implemented in a new test.

**Comment by numerodix** [comment]
@dio All done. See if you have any more feedback.

Btw the asan and gcc build failures look like CI instability and have been haunting other PRs in progress:

asan:
> 2020-09-23T06:50:10.4397266Z ERROR: /source/test/common/http/BUILD:194:14: C++ compilation of rule '//test/common/http:conn_manager_impl_test' failed (Killed): clang failed: error executing command 

gcc:
> 2020-09-23T07:01:32.5177393Z ERROR: /source/test/common/http/BUILD:194:14: C++ compilation of rule '//test/common/http:conn_manager_impl_test' failed (Exit 1): gcc failed: error executing command

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (numerodix) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:350]: I'm assuming because we are iterating over the headers `entry` can never be null here.

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:354]: nit. Sorry, can we just do `headers_to_remove.reserve(header_names.size());` here?

Comment (numerodix) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:354]: Yes, but the header could be present more than once, so we will hit the `HeaderEntry` more than once:
1. The first time we'll do `reserve(0 + header_names.size())`. 
2. The second time we'll do `reserve(<previous header_name.size()> + header_name.size())`

**Review by dio — DISMISSED** [review]
Thanks! Awesome work.

**Comment by numerodix** [comment]
@dio thanks for all the feedback! 
Not sure who has the ability to merge this. @htuch?

**Comment by htuch** [comment]
I just kicked CI again, but you might want to merge master to pick up #13257

**Comment by numerodix** [comment]
@htuch, thanks that finally fixed the Ci jobs

File: source/common/http/headers.h
Comment (mattklein123) [inline source/common/http/headers.h:137]: Can you move these to be defined inside the extauth filter as these are extension specific? Also, where are these magic headers documented? It wasn't clear to me from the docs how I would use these.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Neat. One small comment from a drive by, thank you!

/wait

File: source/common/http/headers.h
Comment (numerodix) [reply source/common/http/headers.h]: @mattklein123 I've moved the two auth headers to a new class in the ext_authz namespace that mimicks how we do things in `headers.h`. I hope that looks okay.

When it comes to documentation the use cases for these headers are a bit obscure. `x-envoy-auth-partial-body` is only used when `with_request_body` is set, and as a result (I imagine) it's only mentioned in that proto field docstring: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto

The new `x-envoy-auth-headers-to-remove` is only used by the http auth server when it's sending a response. It will appear under the new proto field `headers_to_remove` once the PR is merged: https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto.html?highlight=checkrequest#envoy-v3-api-msg-service-auth-v3-okhttpresponse

Do you think these headers should be mentioned in a more central location?

Comment (mattklein123) [reply source/common/http/headers.h]: > The new x-envoy-auth-headers-to-remove is only used by the http auth server when it's sending a response. It will appear under the new proto field headers_to_remove once the PR is merged:

Apologies, I missed this. LGTM!

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
